File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        #把 inorder 变成 值 -> 下标
        indices = {val : idx for idx, val in enumerate(inorder)}
        #指向当前要用的根节点
        self.pre_idx = 0
        
        #构建l, r范围内的树
        def dfs(l, r):
            #如果范围空，则终止
            if l > r:
                return None
            #记录当前节点的值
            root_val = preorder[self.pre_idx]
            self.pre_idx += 1
           #创建新节点
            root = TreeNode(root_val)
            
            #用 inorder 找到左右子树边界
            mid = indices[root_val]
            root.left = dfs(l, mid - 1)
            root.right = dfs(mid + 1, r)
            return root
        return dfs(0, len(inorder) - 1)

File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-1.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    int ans;
    int times;
    public int kthSmallest(TreeNode root, int k) {
        times = k;
        dfs(root);
        return ans;
    }
    private void dfs(TreeNode root) {
        if (times == 0) return;
        if(root == null) return;
        dfs(root.left);
        times -= 1;
        if(times == 0) ans = root.val;
        dfs(root.right);
    }
}
